features/map_test.go

```diff
@@ -0,0 +1,80 @@
+package features
+
+import (
+	"testing"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/internal"
+)
+
+// similar as in testutils, skipTestCaseOnOldKernel helps to determine wether a testcase can be skipped
+// internal.SkipOnOldKernel would skip the entire test which isn't what I want.
+func skipTestCaseOnOldKernel(minVersion string) bool {
+	minV, err := internal.NewVersion(minVersion)
+	if err != nil {
+		panic(err)
+	}
+
+	kVersion, err := internal.KernelVersion()
+	if err != nil {
+		panic(err)
+	}
+
+	if kVersion.Less(minV) {
+		return true
+	}
+
+	return false
+}
+
+func TestHaveMapType(t *testing.T) {
+	type testCase struct {
+		mapType    ebpf.MapType
+		minVersion string
+	}
+
+	testCases := []testCase{
+		{ebpf.Hash, "3.19"},
+		{ebpf.Array, "3.19"},
+		{ebpf.ProgramArray, "4.2"},
+		{ebpf.PerfEventArray, "4.3"},
+		{ebpf.PerCPUHash, "4.6"},
+		{ebpf.PerCPUArray, "4.6"},
+		{ebpf.StackTrace, "4.6"},
+		{ebpf.CGroupArray, "4.8"},
+		{ebpf.LRUHash, "4.10"},
+		{ebpf.LRUCPUHash, "4.10"},
+		{ebpf.LPMTrie, "4.11"},
+		{ebpf.ArrayOfMaps, "4.12"},
+		{ebpf.HashOfMaps, "4.12"},
+		{ebpf.DevMap, "4.14"},
+		{ebpf.SockMap, "4.14"},
+		{ebpf.CPUMap, "4.15"},
+		{ebpf.XSKMap, "4.18"},
+		{ebpf.SockHash, "4.18"},
+		{ebpf.CGroupStorage, "4.19"},
+		{ebpf.ReusePortSockArray, "4.19"},
+		{ebpf.PerCPUCGroupStorage, "4.20"},
+		{ebpf.Queue, "4.20"},
+		{ebpf.Stack, "4.20"},
+		{ebpf.SkStorage, "5.2"},
+		{ebpf.DevMapHash, "5.4"},
+		// {ebpf.StructOpts, "5.6"}, whats the best way to still test this path?
+		{ebpf.RingBuf, "5.8"},
+		{ebpf.InodeStorage, "5.10"},
+		{ebpf.TaskStorage, "5.11"},
+	}
+
+	for _, tc := range testCases {
+		if skipTestCaseOnOldKernel(tc.minVersion) {
+			t.Logf("Skipped test for MapType %s: requires at least kernel version %s", tc.mapType.String(), tc.minVersion)
+			continue
+		}
+
+		err := HaveMapType(tc.mapType)
+		if err != nil {
+			t.Fatalf("MapType %s isn't supported even though kernel is at least %s", tc.mapType.String(), tc.minVersion)
+		}
+	}
+
```

Comment:
This is missing negative tests, the `ErrNotSupported` case is never hit as far as I can see. Try creating a map with mt `math.MaxUint32`, which should fail on all kernel versions.

Tip: Go has built-in coverage reports, see: https://blog.golang.org/cover. (esp. the part about `go tool cover -html=coverage.out`) This should help you build coverage.

Reply:
That test case would only improve the coverage slightly because its caught by the `validateMapType()` check. I could invert `TestHaveMapType` to test new features on old kernels where they don't exist yet to provoke a proper EINVAL and resulting ebpf.ErrNotSupported?

Reply:
Ah, good point. To make this more testable, the current `HaveMapType` could become an exported wrapper around `validateMapType()` and a new `haveMapType()`. This would allow probing to take place from tests without input validation.

>  I could invert TestHaveMapType to test new features on old kernels where they don't exist yet 

That would be the alternative, but as we remove old kernel versions over time, these tests will stop being executed at some point in the future if all kernels we test against are too new.

Reply:
Ok, I'll go with the wrapper approach you are suggesting. This will also fix the currently failing tests.